File: server/internal/models/recipe.go
// Package models defines the JSON-serializable types exchanged between
// the server and the client. They are also the on-disk shape (modulo
// JSON-encoded columns) of the "recipes" table.
package models

import "time"

// Ingredient is one structured row of an ingredient list. We normalize
// TheMealDB's flat strIngredientN/strMeasureN pairs into this struct on
// save so the server stores its own structured copy.
type Ingredient struct {
	Name     string `json:"name"`
	Quantity string `json:"quantity,omitempty"`
	Unit     string `json:"unit,omitempty"`
}

// Recipe is the saved-collection record. `id` is a server-generated
// UUID. `source` + `sourceId` exist so we can de-duplicate when the same
// TheMealDB meal is saved twice.
type Recipe struct {
	ID               string       `json:"id"`
	Title            string       `json:"title"`
	Source           string       `json:"source"` // "themealdb" | "manual"
	SourceID         string       `json:"sourceId,omitempty"`
	Category         string       `json:"category,omitempty"`
	Area             string       `json:"area,omitempty"`
	Tags             []string     `json:"tags"`
	ImageURL         string       `json:"imageUrl,omitempty"`
	PrepTimeMinutes  *int         `json:"prepTimeMinutes,omitempty"`
	CookTimeMinutes  *int         `json:"cookTimeMinutes,omitempty"`
	Servings         *int         `json:"servings,omitempty"`
	Ingredients      []Ingredient `json:"ingredients"`
	Steps            []string     `json:"steps"`
	YouTubeURL       string       `json:"youtubeUrl,omitempty"`
	SourceURL        string       `json:"sourceUrl,omitempty"`
	CreatedAt        time.Time    `json:"createdAt"`
	UpdatedAt        time.Time    `json:"updatedAt"`
}

// RecipeInput is the payload accepted by POST/PUT. We don't trust the
// client-supplied id; the server generates and owns it.
type RecipeInput struct {
	Title            string       `json:"title"`
	Source           string       `json:"source,omitempty"`
	SourceID         string       `json:"sourceId,omitempty"`
	Category         string       `json:"category,omitempty"`
	Area             string       `json:"area,omitempty"`
	Tags             []string     `json:"tags,omitempty"`
	ImageURL         string       `json:"imageUrl,omitempty"`
	PrepTimeMinutes  *int         `json:"prepTimeMinutes,omitempty"`
	CookTimeMinutes  *int         `json:"cookTimeMinutes,omitempty"`
	Servings         *int         `json:"servings,omitempty"`
	Ingredients      []Ingredient `json:"ingredients,omitempty"`
	Steps            []string     `json:"steps,omitempty"`
	YouTubeURL       string       `json:"youtubeUrl,omitempty"`
	SourceURL        string       `json:"sourceUrl,omitempty"`
}

// ErrorResponse is the canonical JSON error shape returned by the API.
// All non-2xx responses include this body.
type ErrorResponse struct {
	Error   string `json:"error"`
	Message string `json:"message,omitempty"`
}

File: client/src/types/recipe.ts
// Shared client-side types. The "Recipe"/"Ingredient" shapes mirror
// the server's models exactly; "MealdbMeal" matches what the server
// proxy forwards from TheMealDB so the proxy responses can be typed.

export interface Ingredient {
  name: string;
  quantity?: string;
  unit?: string;
}

export interface Recipe {
  id: string;
  title: string;
  source: "themealdb" | "manual" | string;
  sourceId?: string;
  category?: string;
  area?: string;
  tags: string[];
  imageUrl?: string;
  prepTimeMinutes?: number | null;
  cookTimeMinutes?: number | null;
  servings?: number | null;
  ingredients: Ingredient[];
  steps: string[];
  youtubeUrl?: string;
  sourceUrl?: string;
  createdAt: string;
  updatedAt: string;
}

export type RecipeInput = Partial<
  Omit<Recipe, "id" | "createdAt" | "updatedAt">
> & { title: string };

/** A single TheMealDB meal record (unaltered upstream shape). */
export interface MealdbMeal {
  idMeal: string;
  strMeal: string;
  strCategory?: string;
  strArea?: string;
  strTags?: string;
  strMealThumb?: string;
  strYoutube?: string;
  strSource?: string;
  strInstructions?: string;
  // strIngredient1..20 / strMeasure1..20 are too noisy to type — read dynamically.
  [key: string]: string | undefined;
}

export interface MealsEnvelope {
  meals: MealdbMeal[] | null;
}

export interface MealdbCategory {
  idCategory: string;
  strCategory: string;
  strCategoryThumb: string;
  strCategoryDescription: string;
}
export interface CategoriesEnvelope {
  categories: MealdbCategory[];
}

export interface MealdbFilterItem {
  idMeal: string;
  strMeal: string;
  strMealThumb: string;
}
export interface FilterEnvelope {
  meals: MealdbFilterItem[] | null;
}
